### ISSUE TITLE
[RFC]: Add GPU ModelRunnerV2 Support for AFD

### ISSUE THREAD

**Issue by jiaran-king** [vllm-project/afd-plugin#223]
| Field | Value |
| --- | --- |
| Status | Draft |
| Upstream | vLLM `v0.26.0` |
| Scope | GPU, synchronous `P2pNcclAFDConnector`, no DBO |
| Related | #171 |

## Motivation

The current AFD GPU path is built on ModelRunnerV1.

This RFC proposes adding ModelRunnerV2 support without copying or rewriting its core execution logic. The migration should preserve existing AFD contracts while keeping native vLLM responsible for request execution, input preparation, Attention, KV cache, CUDA Graph management, sampling, and output construction.

The existing ModelRunnerV1 path will remain available during the migration.

## Proposal

Only the request-facing Attention role will adopt the native ModelRunnerV2 execution architecture.

The FFN role will remain plugin-owned and connector-driven. It will continue to use `AFDFFNWorker` and `GPUFFNModelRunner` without adopting native ModelRunnerV2 request state.

The concrete Attention integration seam will be selected by the Phase 0 proof of concept:

1. a thin `AFDAttentionModelRunnerV2`; or
2. the native GPU ModelRunnerV2 with a narrow hook or compatibility patch.

The RFC does not require a particular class structure. The selected solution must preserve native ModelRunnerV2 ownership and must not copy its execution loop.

## Initial Scope

The first mergeable ModelRunnerV2 path will support:

- CUDA;
- vLLM `v0.26.0`;
- DeepSeek-V2-Lite;
- `1A1F`;
- DP=1, TP=1, static EP;
- synchronous `P2pNcclAFDConnector`;
- `compute_gate_on_attention=false`;
- eager execution;
- prefill and decode;
- deterministic sampling with `temperature=0`;
- DBO disabled.

## Planned No-DBO Parity

After the initial eager path is validated, the migration will restore selected GPU V1 capabilities incrementally:

- `2A2F`, DP=2, TP=1, static EP, eager;
- `2A2F`, DP=1, TP=2, static EP, eager;
- `FULL_DECODE_ONLY` CUDA Graph;
- Attention and FFN profiling;
- OpenAI-compatible serving;
- usage accounting;
- deterministic eager/graph comparison;
- GSM8K accuracy validation.

These capabilities are follow-up milestones, not part of the initial Phase 0 delivery.

## Non-Goals

This RFC does not add or claim support for:

- DBO or native ubatching;
- NPU ModelRunnerV2;
- `compute_gate_on_attention=true`;
- speculative decoding;
- pipeline parallelism;
- sequence-parallel MoE;
- EPLB or elastic EP;
- GPU asynchronous AFD connectors;
- FFN-side LoRA mutation;
- generic GPU quantization;
- generic `torch.compile`;
- CUDA Graph modes other than `FULL_DECODE_ONLY`;
- cross-node NCCL;
- unvalidated model architectures.

Native ModelRunnerV2 support for a feature does not automatically make that feature supported by AFD.

## Architecture

```text
                         Native vLLM Scheduler
                                  |
                                  | SchedulerOutput
                                  v
                         AFDAttentionWorker
                                  |
                                  v
                    Native ModelRunnerV2 ownership
                                  |
          +-----------------------+-----------------------+
          |                       |                       |
          v                       v                       v
     RequestState             InputBatch       ModelCudaGraphManager
          |                       |                       |
          +-----------------------+-----------------------+
                                  |
                      native input/Attention prepare
                                  |
                                  v
                      smallest proven AFD seam
                   thin runner OR hook/compat patch
                                  |
                      build existing AFD metadata
                      send AFDControlPayload
                      install afd_metadata
                                  |
                                  v
                     existing role-aware model
                                  |
                           RemoteFFNProxy
                                  |
                       P2pNcclAFDConnector
                                  |
                                  v
                           AFDFFNWorker
                            daemon loop
                                  |
                                  v
                        GPUFFNModelRunner
                                  |
                            gate + MoE
                                  |
                                  v
                       result returned to Attention
                                  |
                                  v
                  native MRV2 sampling and output
```

## Ownership

Native ModelRunnerV2 remains responsible for:

- `SchedulerOutput` consumption;
- persistent request state;
- `InputBatch`;
- input IDs, positions, and padding;
- Attention metadata;
- block tables and KV cache;
- execution and graph dispatch;
- native graph descriptors and graph storage;
- logits, sampling, and output construction.

AFD remains responsible for:

- role-specific configuration;
- AFD model registration and weight filtering;
- connector and profiler lifecycle;
- `AFDForwardContextMetadata`;
- `AFDControlPayload`;
- control-plane-before-data-plane ordering;
- the remote FFN model boundary;
- rank topology;
- the FFN daemon and FFN computation;
- AFD cleanup.

The existing role-aware DeepSeek model and `RemoteFFNProxy` will continue to consume:

```python
ForwardContext.additional_kwargs["afd_metadata"]
```

No ModelRunnerV2-specific branch should be added to the model layer unless a concrete incompatibility is found.

## V1 Reuse

The following existing components should be reused unless ModelRunnerV2 introduces a concrete incompatibility:

- `P2pNcclAFDConnector`;
- graph-capturable P2P custom operations;
- `AFDControlPayload`;
- `AFDForwardContextMetadata`;
- `AFDTransferMetadata`;
- `RemoteFFNProxy`;
- the role-aware DeepSeek model split;
- rank topology;
- the FFN daemon;
- `GPUFFNModelRunner`;
- the FFN CUDA Graph cache;
- `make_ffn_graph_key()`;
- `graph_run_mode()`;
- profiler helpers.

The behavior represented by the V1 `_model_forward()`, `_dummy_run()`, and `_warmup_and_capture()` overrides must be preserved, but their implementations must not be copied.

All DBO and ubatching-specific code is excluded from the V2 migration.

## Attention Integration Seam

Phase 0 must identify the smallest seam that supports:

- direct role-aware Worker/runner construction;
- eager metadata injection;
- control-plane publication before the first remote FFN transfer;
- dummy and profile execution;
- cleanup after success or failure;
- future per-descriptor graph warmup and capture;
- future per-execution full-graph replay.

A thin runner is acceptable only when the required event matches a narrow upstream method boundary.

If an event is hidden inside a large native method, the implementation should prefer a generic upstream hook or an exact-version compatibility patch.

The implementation must not copy:

- ModelRunnerV2 `execute_model()`;
- native input preparation;
- native Attention preparation;
- `ModelCudaGraphManager.capture()`;
- the native full-graph execution branch.

## Worker Construction

The Attention Worker must construct the selected ModelRunnerV2 integration directly. It must not construct a complete native V2 runner and then discard it.

The FFN Worker must initialize GPU and distributed state without constructing an unused native V1 or V2 runner.

The long-term FFN path should be independent of native ModelRunnerV1/ModelRunnerV2 selection.

## Token-Shape Contract

AFD must distinguish three token-shape concepts:

```text
tokens_unpadded_lens
    = real scheduled tokens

tokens_lens
    = leading dimension of the tensors actually transferred by AFD

dp_metadata_list
    = native DP execution shape after ModelRunnerV2 coordination
```

AFD must not independently repeat DP dispatch or CUDA Graph padding decisions.

For example:

```text
real scheduled tokens = 27
FULL graph execution shape = 32

tokens_unpadded_lens = [27]
tokens_lens          = [32]
```

The FFN side must prepare for the actual transferred shape.

## CUDA Graph

CUDA Graph is not part of the initial eager delivery.

The follow-up `FULL_DECODE_ONLY` path will retain the existing AFD design:

```text
Attention GPU                    FFN GPU

native Attention graph           existing FFN graph
Attention                        recv
P2P send  -------------------->  FFN
P2P recv  <--------------------  send
```

No new graph coordinator or graph protocol will be introduced unless the existing protocol is proven insufficient.

The Attention side requires three kinds of events:

```text
for each capture descriptor:
    before warmup
    before capture

for each FULL execution:
    before replay
```

Each event will send the existing `AFDControlPayload`.

Control-plane communication must remain outside CUDA Graph capture and replay bodies.

### Phase 0: Eager Seam Proof

Implement and validate:

```text
DeepSeek-V2-Lite
1A1F
DP=1
TP=1
static EP
P2pNcclAFDConnector
eager
no DBO
```

Required evidence:

- role-aware Worker/runner construction;
- native ModelRunnerV2 execution;
- AFD metadata and control ordering;
- remote FFN round trip;
- prefill and decode;
- deterministic output;
- repeated requests;
- dummy/profile execution;
- failure and shutdown cleanup;
- no regression to the existing V1 path.

Phase 0 also selects the final Attention seam.

### Phase 1: Selected No-DBO Parity

Add and validate:

- DP=2, TP=1 eager;
- DP=1, TP=2 eager;
- `FULL_DECODE_ONLY`;
- profiling;
- OpenAI-compatible serving;
- usage accounting;
- eager/graph output parity;
- GSM8K accuracy.

### Phase 2: Secondary Coverage

Evaluate separately:

- `4A4F`, DP=2, TP=2;
- NIXL disaggregation;
- other registered AFD models;
- additional valid Attention-to-FFN rank ratios.

## Acceptance Criteria

Phase 0 is complete when:

- Attention executes through native ModelRunnerV2 state machinery;
- the selected AFD seam does not copy the native execution loop;
- the existing connector, metadata, model split, and FFN daemon are reused;
- FFN does not depend on native V1 request execution;
- `1A1F`, DP=1, TP=1 eager prefill/decode is correct;
- native logits, sampling, and output construction remain unchanged;
- DBO and unsupported execution modes fail closed;
- repeated requests and shutdown leave no stale AFD state;
- the existing ModelRunnerV1 path remains functional.

Selected no-DBO parity additionally requires:

- DP and TP eager validation;
- `FULL_DECODE_ONLY` validation;
- serving and profiling regression coverage;
- GSM8K accuracy evidence;
- V1/V2 correctness and performance comparison.

## Review Constraints

An implementation must be rejected if it:

- copies or substantially rewrites ModelRunnerV2 `execute_model()`;
- copies native input, Attention, sampling, or graph-management logic;
- creates AFD-owned `RequestState` or `InputBatch`;
- introduces unnecessary Runtime, Adapter, Coordinator, or Descriptor abstractions;
- constructs and discards an unused native runner on either role;
- ports V1 DBO or ubatching logic into the V2 path;
- executes control-plane operations inside a CUDA Graph;
- collapses real and padded token shapes into one value;
- expands the support claim without corresponding validation evidence.
